File: core/collector.py
from __future__ import annotations

"""
core/collector.py — Orchestrates bridge collection from all sources.

Runs all enabled sources concurrently and merges results into
HistoryManager, deduplicating across transports and IP versions.

Sources (in priority order):
  1. torproject   — bridges.torproject.org scraper
  2. moat         — MOAT/BridgeDB REST API (direct + domain-fronted)
  3. bridgedb_api — BridgeDB HTTPS API (direct + Fastly-fronted)
  4. telegram     — Public Tor-bridge Telegram channels
  5. static       — Hard-coded fallback bridges

Phase 4.3 (ADDITIVE): ``prioritize_port_443`` filter is applied after
scoring so port-443 (HTTPS-disguised) bridges always float to the front
of any bridge list. Iran almost never blocks port 443.
"""


import asyncio
import logging
from typing import Any

import config
from core.history import HistoryManager

log = logging.getLogger(__name__)


def prioritize_port_443(bridges: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """
    Move port-443 bridges to the front of ``bridges``.

    Iran almost never blocks HTTPS port 443, so bridges listening on 443
    get a structural advantage. The relative order of port-443 bridges
    and the relative order of non-443 bridges are both preserved (stable
    partition).

    This filter is ADDITIVE: callers that don't use it see no change.
    """
    p443 = [b for b in bridges if _port_of(b) == 443]
    other = [b for b in bridges if _port_of(b) != 443]
    return p443 + other


def _port_of(bridge: dict[str, Any]) -> int:
    """Best-effort extraction of a bridge's port (0 if unknown)."""
    try:
        return int(bridge.get("port") or 0)
    except Exception:
        return 0


class BridgeCollector:
    def __init__(self, history: HistoryManager):
        self._history = history

    async def collect_all(self) -> int:
        """
        Fetch from all enabled sources concurrently.
        Returns the number of new bridges added to history.
        """
        tasks: list[asyncio.Task] = []

        if config.USE_TORPROJECT_SCRAPER:
            from sources.torproject import fetch_all as tp_fetch
            tasks.append(asyncio.create_task(tp_fetch(), name="torproject"))

        if config.USE_MOAT_API:
            from sources.moat import fetch_all as moat_fetch
            tasks.append(asyncio.create_task(moat_fetch(), name="moat"))

        if config.USE_BRIDGEDB_API:
            from sources.bridgedb_api import fetch_all as bdb_fetch
            tasks.append(asyncio.create_task(bdb_fetch(), name="bridgedb_api"))

        if config.USE_TELEGRAM_SOURCES:
            from sources.telegram_bridges import fetch_all as tg_fetch
            tasks.append(asyncio.create_task(tg_fetch(), name="telegram"))

        if config.USE_GITHUB_SOURCES:
            try:
                from sources.github_bridges import fetch_all as gh_fetch
                tasks.append(asyncio.create_task(gh_fetch(), name="github"))
            except ImportError as _remediation_exc:
                from monitoring.structured_logger import record_silent_failure
                record_silent_failure('core.collector:87', _remediation_exc)
                log.warning("GitHub bridges source not available (import error)")

        gathered: list[list[tuple[str, str, str]]] = await asyncio.gather(
            *tasks, return_exceptions=True
        )

        before = len(self._history.get_all())
        for result in gathered:
            if isinstance(result, Exception):
                log.error("Source error: %s", result)
                continue
            for line, transport, _ip_ver in result:
                self._history.add_bridge(line, transport)

        # Static bridges (synchronous, always run last as fallback)
        if config.USE_STATIC_BRIDGES:
            from sources.static_bridges import get_all
            for line, transport, _ip_ver in get_all():
                self._history.add_bridge(line, transport)

        after = len(self._history.get_all())
        new_count = after - before
        log.info(
            "Collection complete: %d new bridges added (total: %d).",
            new_count, after,
        )
        return new_count

File: src/collector.rs
//! Parity port of `core/collector.py`.
//!
//! Orchestrates bridge collection from all sources. The Python original
//! runs all enabled sources concurrently via `asyncio.gather` and merges
//! results into `HistoryManager`, deduplicating across transports and IP
//! versions.
//!
//! The Rust port exposes the pure decision logic (`prioritize_port_443`,
//! `_port_of`) and a synchronous `BridgeCollector` that accepts an
//! injectable source-fetcher trait. Production callers can wrap the
//! fetcher in their own async runtime.

use serde_json::Value;

use crate::history::HistoryManager;

/// Mirror of `_port_of(bridge)`. Best-effort extraction of a bridge's port.
/// Returns 0 if unknown.
pub fn port_of(bridge: &Value) -> i64 {
    bridge
        .get("port")
        .and_then(|v| {
            v.as_i64()
                .or_else(|| v.as_str().and_then(|s| s.parse::<i64>().ok()))
        })
        .unwrap_or(0)
}

/// Mirror of `prioritize_port_443(bridges)`. Moves port-443 bridges to the
/// front. Stable partition — relative order of port-443 bridges and the
/// relative order of non-443 bridges are both preserved.
pub fn prioritize_port_443(bridges: &[Value]) -> Vec<Value> {
    let mut p443: Vec<Value> = Vec::new();
    let mut other: Vec<Value> = Vec::new();
    for b in bridges {
        if port_of(b) == 443 {
            p443.push(b.clone());
        } else {
            other.push(b.clone());
        }
    }
    p443.extend(other);
    p443
}

/// Trait abstracting a bridge source's `fetch_all()` function. Each source
/// returns a list of `(line, transport, ip_version)` tuples.
pub trait BridgeSource: Sync {
    fn fetch_all(&self) -> Result<Vec<(String, String, String)>, String>;
}

/// Mirror of `BridgeCollector`. Accepts a list of source-fetchers (already
/// filtered by the caller based on `config.USE_*` flags) and merges results
/// into the `HistoryManager`.
pub struct BridgeCollector<'a> {
    history: &'a mut HistoryManager,
    sources: Vec<Box<dyn BridgeSource>>,
}

impl<'a> BridgeCollector<'a> {
    pub fn new(history: &'a mut HistoryManager, sources: Vec<Box<dyn BridgeSource>>) -> Self {
        Self { history, sources }
    }

    /// Mirror of `collect_all()`. Returns the number of new bridges added
    /// to history. Errors from individual sources are logged and skipped
    /// (mirrors Python's `return_exceptions=True` + per-result `isinstance`
    /// check).
    pub fn collect_all(&mut self) -> usize {
        let before = self.history.get_all().len();

        for source in &self.sources {
            match source.fetch_all() {
                Ok(results) => {
                    for (line, transport, _ip_ver) in results {
                        self.history.add_bridge(&line, &transport);
                    }
                }
                Err(e) => {
                    tracing::error!("Source error: {e}");
                }
            }
        }

        let after = self.history.get_all().len();
        let new_count = after.saturating_sub(before);
        tracing::info!("Collection complete: {new_count} new bridges added (total: {after}).");
        new_count
    }
}

/// A trivial source that returns a fixed list. Used in tests.
pub struct StaticSource {
    bridges: Vec<(String, String, String)>,
}

impl StaticSource {
    pub fn new(bridges: Vec<(String, String, String)>) -> Self {
        Self { bridges }
    }
}

impl BridgeSource for StaticSource {
    fn fetch_all(&self) -> Result<Vec<(String, String, String)>, String> {
        Ok(self.bridges.clone())
    }
}

/// A source that always fails. Used in tests to verify error handling.
pub struct FailingSource;

impl BridgeSource for FailingSource {
    fn fetch_all(&self) -> Result<Vec<(String, String, String)>, String> {
        Err("simulated failure".to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::{DateTime, Utc};

    fn fixed_now() -> DateTime<Utc> {
        chrono::DateTime::parse_from_rfc3339("2026-06-28T12:00:00+00:00")
            .unwrap()
            .with_timezone(&Utc)
    }

    #[test]
    fn port_of_extracts_integer_port() {
        let v = serde_json::json!({"port": 443});
        assert_eq!(port_of(&v), 443);
    }

    #[test]
    fn port_of_extracts_string_port() {
        let v = serde_json::json!({"port": "443"});
        assert_eq!(port_of(&v), 443);
    }

    #[test]
    fn port_of_returns_zero_for_missing() {
        let v = serde_json::json!({});
        assert_eq!(port_of(&v), 0);
    }

    #[test]
    fn port_of_returns_zero_for_invalid_string() {
        let v = serde_json::json!({"port": "not-a-number"});
        assert_eq!(port_of(&v), 0);
    }

    #[test]
    fn prioritize_port_443_moves_443_to_front() {
        let bridges = vec![
            serde_json::json!({"port": 9001}),
            serde_json::json!({"port": 443}),
            serde_json::json!({"port": 8080}),
            serde_json::json!({"port": 443}),
        ];
        let result = prioritize_port_443(&bridges);
        // Port 443 bridges come first (in original order), then the rest.
        assert_eq!(port_of(&result[0]), 443);
        assert_eq!(port_of(&result[1]), 443);
        assert_eq!(port_of(&result[2]), 9001);
        assert_eq!(port_of(&result[3]), 8080);
    }

    #[test]
    fn prioritize_port_443_preserves_relative_order() {
        let bridges = vec![
            serde_json::json!({"port": 443, "id": "a"}),
            serde_json::json!({"port": 8080, "id": "b"}),
            serde_json::json!({"port": 443, "id": "c"}),
            serde_json::json!({"port": 9001, "id": "d"}),
        ];
        let result = prioritize_port_443(&bridges);
        // 443 bridges keep their relative order (a before c).
        assert_eq!(result[0]["id"], "a");
        assert_eq!(result[1]["id"], "c");
        // Non-443 bridges keep their relative order (b before d).
        assert_eq!(result[2]["id"], "b");
        assert_eq!(result[3]["id"], "d");
    }

    #[test]
    fn collect_all_adds_bridges_from_all_sources() {
        let dir = std::env::temp_dir().join(format!("collector_test_{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        let mut history = HistoryManager::new(
            &dir.join("hist.json"),
            &dir.join("bridge"),
            &dir.join("export"),
            fixed_now(),
        )
        .unwrap();
        let sources: Vec<Box<dyn BridgeSource>> = vec![
            Box::new(StaticSource::new(vec![
                (
                    "obfs4 1.2.3.4:443".to_string(),
                    "obfs4".to_string(),
                    "ipv4".to_string(),
                ),
                (
                    "snowflake 5.6.7.8:1".to_string(),
                    "snowflake".to_string(),
                    "ipv4".to_string(),
                ),
            ])),
            Box::new(StaticSource::new(vec![(
                "webtunnel 9.10.11.12:443".to_string(),
                "webtunnel".to_string(),
                "ipv4".to_string(),
            )])),
        ];
        let mut collector = BridgeCollector::new(&mut history, sources);
        let added = collector.collect_all();
        assert_eq!(added, 3);
        assert_eq!(history.get_all().len(), 3);
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn collect_all_skips_failing_sources() {
        let dir = std::env::temp_dir().join(format!("collector_test2_{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        let mut history = HistoryManager::new(
            &dir.join("hist.json"),
            &dir.join("bridge"),
            &dir.join("export"),
            fixed_now(),
        )
        .unwrap();
        let sources: Vec<Box<dyn BridgeSource>> = vec![
            Box::new(FailingSource),
            Box::new(StaticSource::new(vec![(
                "obfs4 1.2.3.4:443".to_string(),
                "obfs4".to_string(),
                "ipv4".to_string(),
            )])),
        ];
        let mut collector = BridgeCollector::new(&mut history, sources);
        let added = collector.collect_all();
        assert_eq!(added, 1); // only the second source contributed
        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn collect_all_deduplicates_via_history_manager() {
        let dir = std::env::temp_dir().join(format!("collector_test3_{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        let mut history = HistoryManager::new(
            &dir.join("hist.json"),
            &dir.join("bridge"),
            &dir.join("export"),
            fixed_now(),
        )
        .unwrap();
        // Two sources providing the same bridge line.
        let sources: Vec<Box<dyn BridgeSource>> = vec![
            Box::new(StaticSource::new(vec![(
                "obfs4 1.2.3.4:443".to_string(),
                "obfs4".to_string(),
                "ipv4".to_string(),
            )])),
            Box::new(StaticSource::new(vec![(
                "OBFS4 1.2.3.4:443".to_string(),
                "obfs4".to_string(),
                "ipv4".to_string(),
            )])),
        ];
        let mut collector = BridgeCollector::new(&mut history, sources);
        let added = collector.collect_all();
        assert_eq!(added, 1); // deduplicated by HistoryManager
        let _ = std::fs::remove_dir_all(&dir);
    }
}
